test/cli_test.py

```diff
@@ -348,38 +350,70 @@ def mock_get_pty_info(shell: bool) -> api_pb2.PTYInfo:
 
     async def write_to_fd(fd: int, data: bytes):
         nonlocal captured_out
         captured_out.append((fd, data))
 
+    @contextlib.asynccontextmanager
+    async def fake_stream_from_stdin(handle_input, use_raw_terminal=False):
+        async def _write():
+            message_index = 0
+            while True:
+                if message_index == len(fake_stdin):
+                    break
+                data = fake_stdin[message_index]
+                await handle_input(data, message_index)
+                message_index += 1
+
+        write_task = asyncio.create_task(_write())
+        yield
+        write_task.cancel()
+
     with mock.patch("rich.console.Console.is_terminal", True), mock.patch(
         "modal._pty.get_pty_info", mock_get_pty_info
-    ), mock.patch("modal._container_exec.get_pty_info", mock_get_pty_info), mock.patch(
-        "modal._container_exec.handle_exec_input", asyncnullcontext
-    ), mock.patch("modal._container_exec._write_to_fd", write_to_fd):
-        yield captured_out
+    ), mock.patch("modal.runner.get_pty_info", mock_get_pty_info), mock.patch(
+        "modal._utils.shell_utils.stream_from_stdin", fake_stream_from_stdin
+    ), mock.patch("modal._sandbox_shell.write_to_fd", write_to_fd):
+        yield fake_stdin, captured_out
 
 
 @skip_windows("modal shell is not supported on Windows.")
 def test_shell(servicer, set_env_client, test_dir, mock_shell_pty):
     stub_file = test_dir / "supports" / "app_run_tests" / "default_stub.py"
     webhook_stub_file = test_dir / "supports" / "app_run_tests" / "webhook.py"
+    fake_stdin, captured_out = mock_shell_pty
+
+    fake_stdin.clear()
+    fake_stdin.extend([b'echo "Hello World"\n', b"exit\n"])
 
     # Function is explicitly specified
     _run(["shell", stub_file.as_posix() + "::foo"])
-    assert mock_shell_pty == [(1, b"Hello World")]
-    mock_shell_pty.clear()
+
+    # first captured message is the empty message the mock server sends
+    assert captured_out[1:] == [(1, b"Hello World\n")]
+    captured_out.clear()
 
     # Function is explicitly specified
     _run(["shell", webhook_stub_file.as_posix() + "::foo"])
-    assert mock_shell_pty == [(1, b"Hello World")]
-    mock_shell_pty.clear()
+    assert captured_out[1:] == [(1, b"Hello World\n")]
+    captured_out.clear()
 
     # Function must be inferred
     _run(["shell", webhook_stub_file.as_posix()])
-    assert mock_shell_pty == [(1, b"Hello World")]
-    mock_shell_pty.clear()
+    assert captured_out[1:] == [(1, b"Hello World\n")]
+    captured_out.clear()
+
+
+@skip_windows("modal shell is not supported on Windows.")
+def test_shell_cmd(servicer, set_env_client, test_dir, mock_shell_pty):
+    stub_file = test_dir / "supports" / "app_run_tests" / "default_stub.py"
+    _, captured_out = mock_shell_pty
+    _run(["shell", "--cmd", "pwd", stub_file.as_posix() + "::foo"])
+    p = subprocess.Popen("pwd", stdout=subprocess.PIPE)
+    expected_output = p.stdout.read()
```

Comment:
```suggestion
    expected_output = subprocess.run(["pwd"], capture_output=True, check=True).stdout
```

micro-nit: bit of cold golf to save a line, take it or leave it 👍